server/store/subscriptions.go

```diff
@@ -0,0 +1,200 @@
+package store
+
+import (
+	"encoding/json"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/mattermost/mattermost-plugin-azure-devops/server/constants"
+	"github.com/mattermost/mattermost-plugin-azure-devops/server/serializers"
+)
+
+type SubscriptionStore interface {
+	StoreSubscription(subscription *serializers.SubscriptionDetails) error
+	GetSubscriptionList() (*SubscriptionList, error)
+	GetAllSubscriptions(userID string) ([]*serializers.SubscriptionDetails, error)
+	DeleteSubscription(subscription *serializers.SubscriptionDetails) error
+}
+
+type SubscriptionListMap map[string]serializers.SubscriptionDetails
+
+type SubscriptionList struct {
+	ByMattermostUserID map[string]SubscriptionListMap
+}
+
+func NewSubscriptionList() *SubscriptionList {
+	return &SubscriptionList{
+		ByMattermostUserID: map[string]SubscriptionListMap{},
+	}
+}
+
+func storeSubscriptionAtomicModify(subscription *serializers.SubscriptionDetails, initialBytes []byte) ([]byte, error) {
+	subscriptionList, err := SubscriptionListFromJSON(initialBytes)
+	if err != nil {
+		return nil, err
+	}
+
+	subscriptionList.AddSubscription(subscription.MattermostUserID, subscription)
+	modifiedBytes, marshalErr := json.Marshal(subscriptionList)
+	if marshalErr != nil {
+		return nil, marshalErr
+	}
+	return modifiedBytes, nil
+}
+
+func (s *Store) StoreSubscription(subscription *serializers.SubscriptionDetails) error {
+	key := GetSubscriptionListMapKey()
+	if err := s.AtomicModify(key, func(initialBytes []byte) ([]byte, error) {
+		return storeSubscriptionAtomicModify(subscription, initialBytes)
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (subscriptionList *SubscriptionList) AddSubscription(userID string, subscription *serializers.SubscriptionDetails) {
+	if _, valid := subscriptionList.ByMattermostUserID[userID]; !valid {
+		subscriptionList.ByMattermostUserID[userID] = make(SubscriptionListMap)
+	}
+
+	subscriptionListValue := serializers.SubscriptionDetails{
+		MattermostUserID:                 userID,
+		ProjectName:                      subscription.ProjectName,
+		ProjectID:                        subscription.ProjectID,
+		OrganizationName:                 subscription.OrganizationName,
+		ChannelID:                        subscription.ChannelID,
+		EventType:                        subscription.EventType,
+		ServiceType:                      subscription.ServiceType,
+		SubscriptionID:                   subscription.SubscriptionID,
```

Comment:
If this is an internally generated id, can we name the field `ID`, and put it at the top of the struct? Also, how is this currently being generated @avas27JTG?

Reply:
This ID is generated and returned from the Azure DevOps when we hit their API to create subscriptions
also the field name is `subscriptionID` in the API response so we also kept the same name for the variable

have moved it at the top